src/datatypes/field.rs

```diff
@@ -6,29 +6,47 @@ use crate::{datatypes::dtype::DataType, error::DaftResult};
 
 use serde::{Deserialize, Serialize};
 
+pub type Metadata = std::collections::BTreeMap<String, String>;
+
 #[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize, Hash)]
 pub struct Field {
     pub name: String,
     pub dtype: DataType,
+    pub metadata: Metadata,
```

Comment:
Just to double-check my understanding, the main motivation for shared ownership of `metadata` is to make `Field` cheap to clone even if `metadata` is large, right? If so, I think it would be better to `Arc` things at the abstraction level where the cloning takes place, i.e. keep this struct pure and lift the `Arc`ing to the higher-level data structures that hold a `Field` and that actually do the cloning, since I doubt that `metadata` is going to be cloned independently from the rest of `Field`.

We have already done this with `DataArray.field`, and I think we should do something similar for the `DataType` enum; Arrow2 is actively doing exactly that, actually! https://github.com/jorgecarleitao/arrow2/pull/1469

Reply:
Yes that intention! We should already refer to Field via an `Arc` so clones will be cheap as you mentioned. But  if we do a `rename` or `cast` then it would copy the entire BTree!

Reply:
Ah got it! I was thinking about `Field`/`DataType`/`DataArray`-level clones, not `Field` mutations, that's a good point. I'll update this to an `Arc`.